### ISSUE TITLE
persist: duplicate key value violates unique constraint "pg_type_typname_nsp_index"

### ISSUE THREAD

**Issue by danhhz** [MaterializeInc/materialize#12560]
Philip found the following log message in CI output for an mzcompose test that is starting and stopping replicas running against https://github.com/MaterializeInc/materialize/issues/12549.

```
storaged: 2022-05-18T16:58:57.683272Z ERROR mz_storage::source: Failed to create source kafka-u7 timestamper: error creating persist client: indeterminate: db error: ERROR: duplicate key value violates unique constraint "pg_type_typname_nsp_index"
DETAIL: Key (typname, typnamespace)=(consensus, 2200) already exists.
```

"creating persist client" only appears in one place in the codebase, which is a PersistLocation::open call. From the message, it's clearly coming from postgres, and thus from PostgresConsensus::open. This simple connects to postgres and runs the following in a txn:

```
CREATE TABLE IF NOT EXISTS consensus (
    shard text NOT NULL,
    sequence_number bigint NOT NULL,
    data bytea NOT NULL,
    PRIMARY KEY(shard, sequence_number)
);
```

Philip found some references on stack overflow (including https://stackoverflow.com/questions/54351783/duplicate-key-value-violates-unique-constraint-postgres-error-when-trying-to-c") that this might be expected when trying to run `CREATE ... IF NOT EXISTS` concurrently from several places.

If that's the case, a sufficient fix may to simply to sniff out this error code and retry it within PostgresConsensus::open. I wrote a unit test to stress concurrent usage of PostgresConsensus, including open, but wasn't able to reproduce it. That test seems useful on it's own, so I'll go ahead and PR it, but I'm inclined to otherwise leave this alone until/if we see it again.

**Comment by benesch** [comment]
Ugh, this reproduces 100% on #12544. It's definitely a known bug in PostgreSQL, as you mentioned. Here are two pretty authoritative references:

  * https://www.postgresql.org/message-id/CA%2BTgmoZAdYVtwBfp1FL2sMZbiHCWT4UPrzRLNnX1Nb30Ku3-gg%40mail.gmail.com
  * https://stackoverflow.com/a/29908840

I'm going to push up a commit to #12544 that works around the issue with advisory locks, as suggested in the first link.

**Comment by danhhz** [comment]
Fixed by #12566
